find_email_addresses.py

```diff
@@ -0,0 +1,156 @@
+import sys
+import re
+
+try:
+    # Python 3
+    from urllib.request import urlopen
+    import urllib.parse as parse
+    from urllib.error import URLError
+except ImportError:
+    # Python 2
+    from urllib2 import urlopen, URLError
+    import urlparse as parse
+
+
+def getUrl(domain):
+    """
+    Given a domain, return a valid URL, inserting a scheme if necessary.
+    """
+
+    p = parse.urlparse(domain)
+    if p.scheme == '':
+        p = parse.urlparse('http://' + domain)
+
+    return p.geturl()
+
+
+def findEmailAddresses(domain):
+    """
+    Given a domain name, find and return a list of email addresses on that
+    web page, and any discoverable web page in the domain.
+    """
+
+    # First, construct a valid url from the domain name
+    url = getUrl(domain)
+
+    # Crawl
+    addresses = crawlForEmail(url)
+    return addresses
+
+
+def getPage(url):
+    """
+    Retrieve webpage stream for a given URL. If the requested resource is not
+    found or an error occurs -1 will be returned.  If a timeout error occurs,
+    this will re-attempt opening the page 3 times before giving up.
+    """
+
+    attempts = 0
+
+    while attempts < 3:
+        try:
+            resp = urlopen(url)
+            return resp
+        except URLError, e:
+            # If we get a 408, try again, on any other error exit
+            if hasattr(e, 'code') and e.code == 408:
+                attempts += 1
```

Comment:
The idea was that the 408 would indicate that the resource was temporarily unavailable, so that if we retried a few times it may become available. This might not be what we want because of the server is returning a 408, it's unlikely that we'll get the response by requesting twice more. It is likely better to instead handle timeouts on the opposite end, if the urllib request is timing out retry, but if the server returns an error pass.